deployment doc (#3)
EOEPCA/document-mlops: docs/getting-started/quick-start.md

# Quick Start

> For this deployment you need a Gitlab instance, if you have one or use `Gitlab.com`, you can skip the first part and directly to [Sharinghub](#sharinghub)

Reviewer: ```suggestion
For this deployment you will need a Gitlab instance, if you already have one or use `Gitlab.com` you can skip the first part and go directly to [Sharinghub](#sharinghub).
```

## Gitlab

To deploy this application you need to:

- [Create two S3 bucket](#create-two-s3-bucket)

> - [*Optional* Configure OIDC](#configure-oidc)

Reviewer: ```suggestion
- *Optional* [Configure OIDC](#configure-oidc)
```

- [Deploy the Gitlab ArgoCD Application](#deploy-the-gitlab-argocd-application)

This gitlab deployment comes in minimal mode, with Keycloak OIDC preconfigured.
If you want to customize your Gitlab deployment, you can check this [link](https://docs.gitlab.com/charts/charts/).

#### Create two S3 bucket

Create two S3 bucket `gitlab-backup-storage` and `gitlab-tmp-storage` in your favorite S3 provider.

> Them create a S3 config file called `storage.config`:

Reviewer: ```suggestion
Then, create a S3 config file called `storage.config`:
```

```conf
[default]
access_key = <access_key>
bucket_location = <bucket_region>
host_base = <s3_endpoint>
secret_key = <secret_key>
use_https = True
```

> Finally create the s3 config secret:

Reviewer: ```suggestion
Finally, create the s3 config secret:
```

```bash
kubectl create ns gitlab
kubectl create secret generic storage-config --from-file=config=storage.config -n gitlab
```

For more details about object storage configuration go [here](https://docs.gitlab.com/charts/advanced/external-object-storage/index.html).

#### Configure OIDC

> This step is optional, if you want to connect your Gitlab instance to your Keycloak realm.

Reviewer: ```suggestion
This step is optional, continue if you want to connect your Gitlab instance to your Keycloak realm.
```

> Create a ``gitlab` client in your Keycloak realm.

Reviewer: ```suggestion
Create a `gitlab` client in your Keycloak realm.
```

            backups:
              objectStorage:
                config:
                  secret: storage-config
                  key: config

          sidekiq:
            enabled: true

        registry:
          enabled: false
          ingress:
            tls:
              secretName: registry.<domain_name>-tls

        minio:
          ingress:
            tls:
              secretName: minio.<domain_name>-tls

        certmanager-issuer:
          email: me@example.com

        certmanager:
          install: false

        nginx-ingress:
          enabled: false

        nginx-ingress-geo:
          enabled: false

        prometheus:
          install: false

        gitlab-runner:
          install: false

        redis:
          metrics:
            enabled: false

        postgresql:
          metrics:
            enabled: false

  syncPolicy:
    syncOptions:
      - FailOnSharedResource=true
      - CreateNamespace=true
```

Before you deploy the `gitlab.yaml` application, you will need to fill some fields:

- `<domain_name>` with your domain name (ex: example.com).

Finally, apply the Application manifest:

```bash
kubectl apply -f gitlab.yaml
```

> For more detail about Gitlab configuration their [doc](https://docs.gitlab.com/charts/)

Reviewer: ```suggestion
For more detail about Gitlab configuration check their [documentation](https://docs.gitlab.com/charts/).
```

## Sharinghub

> The Helm charts are in [SharingHub repo](https://github.com/csgroup-oss/sharinghub) in `deploy/helm/sharinghub`.

Reviewer: ```suggestion
The Helm charts are in the [SharingHub repository](https://github.com/csgroup-oss/sharinghub), in the folder `deploy/helm/sharinghub`.
```

```txt
http://localhost:8000/api/auth/login/callback
https://sharinghub.<domain-name>/api/auth/login/callback
```

> Note: localhost URL is for development purposes, if you don't want it you can remove it.

![Configure application](../img/configure-application.png)

You will be given a secret, and an application id, we will use them to create the following secret:

```bash
kubectl create secret generic sharinghub-oidc --from-literal client-id="<application-id>" --from-literal client-secret="<application-secret>" --namespace sharinghub
```

If you want, you can also set a default token, allowing use to use SharingHub without being authenticated, because the token will be used as default. The default token can be a personal Access Token, or a Group Access Token. Group tokens needs at least the role `Reporter`, and the scopes `read_api`, `read_repository`. Personal tokens needs `api` scope.

The secret creation will look like this:

```bash
kubectl create secret generic sharinghub-oidc  --from-literal default-token="<default-token>" --from-literal client-id="<client-id>" --from-literal client-secret="<client-secret>" --namespace sharinghub
```

### Deploy Sharinghub

> Next, create a file called `sharinghub.yaml`:

Reviewer: ```suggestion
Next, create a file called `sharinghub.yaml`:

```

          pullPolicy: IfNotPresent
          # Overrides the image tag whose default is the chart appVersion.
          tag: "latest"

        ingress:
          enabled: true
          className: "nginx"
          annotations:
            cert-manager.io/cluster-issuer: letsencrypt-prod
            nginx.ingress.kubernetes.io/ssl-redirect: "true"
            nginx.ingress.kubernetes.io/proxy-body-size: 10g
          hosts:
            - host: sharinghub.<domain_name>
              paths:
                - path: /
                  pathType: ImplementationSpecific
          tls:
            - secretName: sharinghub.<domain_name>-tls
              hosts:
                - sharinghub.<domain_name>

  syncPolicy:
    syncOptions:
      - FailOnSharedResource=true
      - CreateNamespace=true
    automated:
      selfHeal: true
```

You will need to fill some fields:

- `<domain_name>` with your platform domain name (ex: example.com).

And, if you created an S3 bucket, complete the configuration (`spec.source.helm.valuesObject.config`).

```yaml
config: |-
  s3:
    enable: true
    bucket: <s3-bucket>
    region: <s3-region>
    endpoint: https://<s3-endpoint>
```

> 

Reviewer: ```suggestion

If you want to learn more about the configuration, please check the server [CONFIGURATION.md](https://github.com/csgroup-oss/sharinghub-server/blob/main/CONFIGURATION.md).

```
